storage: trigger Raft log truncation based on size, not entry count

petermattis opened the pull request:
Trigger proactive Raft log truncation based on the size of the Raft log
since the last check. Truncate the Raft log if either the number of
entries or size exceeds a threshold.

Fixes #15210

petermattis committed:
storage: trigger Raft log truncation based on size, not entry count

Trigger proactive Raft log truncation based on the size of the Raft log
since the last check. Truncate the Raft log if either the number of
entries or size exceeds a threshold.

Fixes #15210

petermattis commented:
Cc @danhhz

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15799)
<!-- Reviewable:end -->

tamird commented:
Perhaps I'm missing some context from #15210, but why does this new mechanism depend on the size of the raft log *since the last check* rather than the absolute size of the raft log?

---

Reviewed 1 of 4 files at r1.
Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/raft_log_queue.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhcq9x9dpJUMCyP5Pq:-Kjhcq9x9dpJUMCyP5Pr:b-mfmy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/raft_log_queue.go#L87)):*
> ```Go
> }
> 
> // getTruncatableIndexes returns the number of truncatable indexes and the
> ```

could you update this comment?

---

*[pkg/storage/replica_proposal.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhc-bXnj43-VK_w87H:-Kjhc-bXnj43-VK_w87I:bl3134k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/replica_proposal.go#L670)):*
> ```Go
> 			r.mu.raftLogSize = 0
> 		}
> 		if r.mu.raftLogLastCheckSize > r.mu.raftLogSize {
> ```

how come this is conditional?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-:-KjhdDRMJo1za5gHV0rF:bcarzuy)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@tamird In the event that we enqueue the range and there are no truncatable entries, we want to wait a bit before trying again instead of re-enqueuing on every command once we reach this size. 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica_proposal.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhc-bXnj43-VK_w87H:-Kjhu21zbx0NONEWhpSP:b-aj0e1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/replica_proposal.go#L670)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come this is conditional?
</blockquote></details>

Yeah, it looks like this might be better as an unconditional `r.mu.raftLogLastCheckSize += *rResult.RaftLogDelta` (and maybe the same clamp-to-zero logic)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-:-KjhuAUY46HNxeLw2E8R:b-bymg5e)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@tamird What Ben said. I added a comment near the call to `raftLogQueue.MaybeAdd`.

---

Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica_proposal.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhc-bXnj43-VK_w87H:-Kji4EWQuhOALYvSx3O8:brsc1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/replica_proposal.go#L670)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, it looks like this might be better as an unconditional `r.mu.raftLogLastCheckSize += *rResult.RaftLogDelta` (and maybe the same clamp-to-zero logic)
</blockquote></details>

Ok, changed. Note that `*RaftLogDelta` is always negative as it is only set in `evalTruncateLog`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-:-Kji59bOAGPoF576Z3_p:bn9w69o)*
<!-- Sent from Reviewable.io -->

tamird commented:
Thanks for adding the comment! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/raft_log_queue.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji4tUDYkyw7qqTnB4I:-Kji4tUDYkyw7qqTnB4J:bsiue3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/raft_log_queue.go#L46)):*
> even if there are fewer entries to truncate

Fewer entries than what?

---

*[pkg/storage/replica.go, line 291 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji67HB0Ybx9tCT0oTR:-Kji67HB0Ybx9tCT0oTS:b-ah2394) ([raw file](https://github.com/cockroachdb/cockroach/blob/c44c3a1cb226e609de39230c8d20ee7172d53888/pkg/storage/replica.go#L291)):*
> ```Go
> 		// raftLogLastCheckSize is the value of raftLogSize the last time the Raft
> 		// log was checked for truncation.
> 		raftLogLastCheckSize int64
> ```

I think this name (and comment) is a bit confusing because this field is also updated on truncation, so it's really the value of raftLogSize at last check or after last truncation, whichever was later. Could you update the comment and/or rename the field to clarify that?

---

*[pkg/storage/replica_proposal.go, line 670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhc-bXnj43-VK_w87H:-Kji63_fbjVgxBsUMUer:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/replica_proposal.go#L670)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, changed. Note that `*RaftLogDelta` is always negative as it is only set in `evalTruncateLog`. 
</blockquote></details>

Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-:-Kji48yYc0x5q8-B14WE:byd6uit)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/raft_log_queue.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji4tUDYkyw7qqTnB4I:-Kji9P1wazfDyngNW_Ku:b2assv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/raft_log_queue.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> even if there are fewer entries to truncate

Fewer entries than what?
</blockquote></details>

Fewer than `RaftLogQueueStaleThreshold`. Fixed the comment.

---

*[pkg/storage/raft_log_queue.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhcq9x9dpJUMCyP5Pq:-Kji9yK-jbQtjGMkk2vE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/raft_log_queue.go#L87)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

could you update this comment?
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 291 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji67HB0Ybx9tCT0oTR:-KjiAhPHdfAXHT9cU8is:b-f2ky6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c44c3a1cb226e609de39230c8d20ee7172d53888/pkg/storage/replica.go#L291)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think this name (and comment) is a bit confusing because this field is also updated on truncation, so it's really the value of raftLogSize at last check or after last truncation, whichever was later. Could you update the comment and/or rename the field to clarify that?
</blockquote></details>

I adjusted the comment slightly, but I'm not sure that is adequate. I can't think of a better name. Any suggestions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/raft_log_queue.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kjhcq9x9dpJUMCyP5Pq:-KjiBJCXuZJWbALn-EPU:bargorf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f8af555678063408de90e36c09f6b4f3a22de61/pkg/storage/raft_log_queue.go#L87)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

and thousands mourned the oxford comma.

---

*[pkg/storage/replica.go, line 291 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji67HB0Ybx9tCT0oTR:-KjiDW1qE4vdKDj4QpkS:b-8x8u95) ([raw file](https://github.com/cockroachdb/cockroach/blob/c44c3a1cb226e609de39230c8d20ee7172d53888/pkg/storage/replica.go#L291)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I adjusted the comment slightly, but I'm not sure that is adequate. I can't think of a better name. Any suggestions?
</blockquote></details>

It would be clearer if you flipped the semantics and made this "raftLogNextCheckSize", but that would require some rewriting of this change.

I don't have a better name either, given the current behaviour.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 291 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799#-Kji67HB0Ybx9tCT0oTR:-KjiFJJn7UwkVZXstjFm:b-xlhipn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c44c3a1cb226e609de39230c8d20ee7172d53888/pkg/storage/replica.go#L291)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It would be clearer if you flipped the semantics and made this "raftLogNextCheckSize", but that would require some rewriting of this change.

I don't have a better name either, given the current behaviour.
</blockquote></details>

I'm going to leave as-is. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15799)*
<!-- Sent from Reviewable.io -->
